core/cache/httpFrontend_test.go

```diff
@@ -0,0 +1,281 @@
+	ret := _m.Called(key, entry)
+
+	var r0 error
+	if rf, ok := ret.Get(0).(func(string, *Entry) error); ok {
+		r0 = rf(key, entry)
+	} else {
+		r0 = ret.Error(0)
+	}
+
+	return r0
+}
+
+func createLoader(statusCode int, body string, err error) func(ctx context.Context) (*http.Response, *Meta, error) {
+	return func(ctx context.Context) (*http.Response, *Meta, error) {
+		return createResponse(statusCode, body), nil, err
+	}
+}
+
+func createResponse(statusCode int, body string) *http.Response {
+	response, _ := http.ReadResponse(
+		bufio.NewReader(
+			strings.NewReader(
+				fmt.Sprintf("HTTP/1.1 %d OK\nContent-Type: text/html\n\n%s", statusCode, body),
+			),
+		),
+		nil,
+	)
+	return response
+}
+
+func TestHTTPFrontend_Get(t *testing.T) {
+	// wait channel top check async cache setting
+	cacheSetComplete := make(chan struct{}, 1)
+	inFuture := time.Now().Add(time.Hour)
+	inPast := time.Now().Add(-time.Hour)
+
+	type args struct {
+		key    string
+		loader HTTPLoader
+	}
+	tests := []struct {
+		name string
+		args args
+		// Entry returned by Backend Get
+		cacheEntry *Entry
+		// expected response from HTTPFrontend
+		want    *http.Response
+		wantErr bool
+		// wanted data to be passed to Backend Set
+		wantedCachedData []byte
+		// want Backend Set to be called
+		wantSet bool
+	}{
+		{
+			name: "empty cache",
+			args: args{
+				key:    "test",
+				loader: createLoader(200, "body", nil),
+			},
+			cacheEntry:       nil,
+			want:             createResponse(200, "body"),
+			wantedCachedData: []byte("body"),
+			wantSet:          true,
+			wantErr:          false,
+		},
+		{
+			name: "empty cache, error on loader",
+			args: args{
+				key:    "test",
+				loader: createLoader(200, "body", errors.New("test error on loader")),
+			},
+			cacheEntry: nil,
+			want:       nil,
+			// even in error case of the loader, the result is expected to be cached (easy circuit breaker)
+			wantedCachedData: nil,
+			wantSet:          true,
+			wantErr:          true,
+		},
+		{
+			name: "data from cache in lifetime",
+			args: args{
+				key:    "test",
+				loader: nil,
+			},
+			cacheEntry: &Entry{
+				Meta: Meta{
+					lifetime:  inFuture,
+					gracetime: inFuture,
+				},
+				Data: cachedResponse{
+					orig: createResponse(200, "foo"),
+					body: []byte("foo"),
+				},
+			},
+			want:             createResponse(200, "foo"),
+			wantedCachedData: nil,
+			wantSet:          false,
+			wantErr:          false,
+		},
+		{
+			name: "from cache out of lifetime but in gracetime",
+			args: args{
+				key:    "test",
+				loader: createLoader(200, "body", nil),
+			},
+			cacheEntry: &Entry{
+				Meta: Meta{
+					lifetime:  inPast,
+					gracetime: inFuture,
+				},
+				Data: cachedResponse{
+					orig: createResponse(200, "foo"),
+					body: []byte("foo"),
+				},
+			},
+			// we expect the cached value as result, but a the actual value from loader to be cached (async)
+			want:             createResponse(200, "foo"),
```

Comment:
Isn't this too inconsistent - why wouldn't you want to use the latest response?

Reply:
getting the latest response can take a lot of time. We want to answer quickly with the old response from the cache and assure that the next requests get the new result though